search/doc_vec.hpp

```diff
@@ -0,0 +1,322 @@
+      maxWeights[n - 1] = std::max(maxWeights[n - 1], tws[i].m_weight);
+    }
+  }
+
+  ASSERT_LESS_OR_EQUAL(n, tws.size(), ());
+  tws.erase(tws.begin() + n, tws.end());
+}
+
+// Computes squared L2 norm of vector of tokens.
+template <typename Token>
+double SqrL2(std::vector<TokenWeightPair<Token>> const & tws)
+{
+  double sum = 0;
+  for (auto const & tw : tws)
+    sum += tw.SqrWeight();
+  return sum;
+}
+
+// Computes squared L2 norm of vector of tokens + prefix token.
+template <typename Token>
+double SqrL2(std::vector<TokenWeightPair<Token>> const & tws,
+             boost::optional<TokenWeightPair<Token>> const & prefix)
+{
+  double result = SqrL2(tws);
+  return result + (prefix ? prefix->SqrWeight() : 0);
+}
+}  // namespace impl
+
+// This class represents a document in a vector space of tokens.
+template <typename Token>
+class DocVec
+{
+public:
+  using TokenWeightPair = TokenWeightPair<Token>;
+  using TokenWeightPairs = std::vector<TokenWeightPair>;
+
+  class Builder
+  {
+  public:
+    template <typename T>
+    void Add(T && token, double weight)
+    {
+      m_tws.emplace_back(std::forward<T>(token), weight);
+    }
+
+  private:
+    friend class DocVec;
+
+    TokenWeightPairs m_tws;
+  };
+
+  DocVec() = default;
+  explicit DocVec(Builder && builder) : m_tws(std::move(builder.m_tws)) { Init(); }
+  explicit DocVec(Builder const & builder) : m_tws(builder.m_tws) { Init(); }
+
+  TokenWeightPairs const & GetTokenWeightPairs() const { return m_tws; }
+  std::vector<double> const & GetMaxWeights() const { return m_maxWeights; }
+
+  bool Empty() const { return m_tws.empty(); }
+
+private:
+  template <typename T>
+  friend std::string DebugPrint(DocVec<T> const & dv)
+  {
+    return "DocVec " + DebugPrint(dv.m_tws);
+  }
+
+  void Init() { impl::SortAndMerge(m_tws, m_maxWeights); }
+
+  TokenWeightPairs m_tws;
+  std::vector<double> m_maxWeights;
+};
+
+// This class represents a search query in a vector space of tokens.
+template <typename Token>
+class QueryVec
+{
+public:
+  using TokenWeightPair = TokenWeightPair<Token>;
+  using TokenWeightPairs = std::vector<TokenWeightPair>;
+
+  class Builder
+  {
+  public:
+    template <typename T>
+    void AddFull(T && token, double weight)
+    {
+      m_tws.emplace_back(std::forward<T>(token), weight);
+    }
+
+    template <typename T>
+    void SetPrefix(T && token, double weight)
+    {
+      m_prefix = TokenWeightPair(std::forward<T>(token), weight);
+    }
+
+  private:
+    friend class QueryVec;
+
+    TokenWeightPairs m_tws;
+    boost::optional<TokenWeightPair> m_prefix;
+  };
+
+  QueryVec() = default;
+
+  explicit QueryVec(Builder && builder)
+    : m_tws(std::move(builder.m_tws)), m_prefix(std::move(builder.m_prefix))
+  {
+    Init();
+  }
+
+  explicit QueryVec(Builder const & builder) : m_tws(builder.m_tws), m_prefix(builder.m_prefix)
+  {
+    Init();
+  }
+
+  // Computes cosine distance between |*this| and |rhs|.
```

Comment:
s/distance/similarity/
As far as I can tell, the convention is ```distance = 1 - similarity```

Reply:
OK